### PR TITLE
added examples and py utils

### PR DESCRIPTION
**Pull request by drew-512**

### CODE REVIEW

**Review by sbinet — COMMENTED** [review]
looks good.

but could we find a way to wire the examples into either a `func ExampleXXX` or `func TestXXX` scaffolding?
so everything is tested w/ `go test ./...`?
(I guess `func TestXXX` would be easier)

**Comment by drew-512** [comment]
I had that thought as well.    If I make the example's main optionally output to a file, in the TestXXX variant I could just compare that output against a known output as the test.  How does that sound?

**Comment by drew-512** [comment]
How does this look?   I'm not well versed in this sort of stuff, so lmk what I'm missing, thanks.

**Comment by drew-512** [comment]
@sbinet This works on my setup but it looks like I need help getting it work with CI here.

**Comment by sbinet** [comment]
it probably is comming from this:

```
=== RUN   TestEmbeddedExample
    main_test.go:64: could not run diff command: exit status 1
    main_test.go:66: test output differ:
        --- testdata/embedding_golden.txt	2022-02-12 14:36:59.468987820 +0100
        +++ /tmp/go-python-embedding-181936362/out.txt	2022-02-12 14:50:15.575695331 +0100
        @@ -4,7 +4,7 @@
         
         ==========================================================
                 Python 3.4 (github.com/go-python/gpython)
        -        go1.17.6 on darwin amd64
        +        devel go1.18-a5c0b19080 Sun Jan 30 02:29:51 2022 +0000 on linux amd64
         ==========================================================
         
         Spring Break itinerary:
--- FAIL: TestEmbeddedExample (0.50s)
FAIL
FAIL	github.com/go-python/gpython/examples/embedding	0.507s
FAIL
```

File: examples/embedding/main_test.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package main
+
+import (
+	"bytes"
+	"os"
+	"os/exec"
+	"testing"
+)
+
+const embeddingTestOutput = "embedding-test-output.txt"
```
Comment (sbinet) [inline examples/embedding/main_test.go:10]:
```suggestion
const embeddingTestOutput = "testdata/embedding_golden.txt"
var regen = flag.Bool("regen", false, "regenerate golden files")
```

File: examples/embedding/main_test.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+func TestEmbeddedExample(t *testing.T) {
+
+	cmd := exec.Command("go", "build", ".")
```
Comment (sbinet) [inline examples/embedding/main_test.go:14]:
```suggestion
	tmp, err := os.MkdirTemp("", "go-python-embedding-")
	if err != nil { t.Fatal(err) }
	defer os.RemoveAll(tmp)
	cmd := exec.Command("go", "build", "-o", filepath.Join(tmp,"exe"), ".")
```

File: examples/embedding/main_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	err := cmd.Run()
+	if err != nil {
+		t.Fatalf("failed to compile embedding example: %v", err)
+	}
+
+	out := new(bytes.Buffer)
+	cmd = exec.Command("./embedding", "mylib-demo.py")
```
Comment (sbinet) [inline examples/embedding/main_test.go:21]:
```suggestion
	cmd = exec.Command(filepath.Join(tmp,"exe"), "mylib-demo.py")
```

File: examples/embedding/main_test.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	resetTest := false // true
+	testOutput := out.Bytes()
+	if resetTest {
+		err = os.WriteFile(embeddingTestOutput, testOutput, 0644)
+		if err != nil {
+			t.Fatalf("failed to write test output: %v", err)
+		}
+	} else {
```
Comment (sbinet) [inline examples/embedding/main_test.go:36]:
```suggestion
	testOutput := out.Bytes()
	if *regen {
		err = os.WriteFile(embeddingTestOutput, testOutput, 0644)
		if err != nil {
			t.Fatalf("failed to write test output: %v", err)
		}
	}
```

File: examples/embedding/main_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		mustMatch, err := os.ReadFile(embeddingTestOutput)
+		if err != nil {
+			t.Fatalf("failed read %q", embeddingTestOutput)
+		}
+		if !bytes.Equal(testOutput, mustMatch) {
+			t.Fatalf("embedded test output did not match accepted output from %q", embeddingTestOutput)
+		}
+	}
```
Comment (sbinet) [inline examples/embedding/main_test.go:44]:
```suggestion
```

File: examples/embedding/mylib.module.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	py.RegisterModule(&py.ModuleImpl{
+		Info: py.ModuleInfo{
+			Name: "mylib_go",
+			Doc:  "Example embedded python module",
+		},
+		Methods: methods,
+		Globals: py.StringDict{
+			"PY_VERSION": py.String("Python 3.4 (github.com/go-python/gpython)"),
+			"GO_VERSION": py.String(fmt.Sprintf("%s on %s %s", runtime.Version(), runtime.GOOS, runtime.GOARCH)),
+			"MYLIB_VERS": py.String("Vacaton 1.0 by Fletch F. Fletcher"),
```
Comment (sbinet) [inline examples/embedding/mylib.module.go:48]:
```suggestion
			"MYLIB_VERS": py.String("Vacation 1.0 by Fletch F. Fletcher"),
```

File: examples/embedding/lib/mylib.py [hunk 7]
```diff
@@ -1,15 +1,13 @@
...
-GO_VERSION  = _go.GO_VERSION
 
 
 print('''
 ==========================================================
         %s
-        %s
 ==========================================================
-''' % (PY_VERSION, GO_VERSION))
+''' % (PY_VERSION))
```
Comment (sbinet) [inline examples/embedding/lib/mylib.py:10]: my old python2 sense would say:
```suggestion
''' % (PY_VERSION, ))
```

File: examples/embedding/main_test.go [hunk 8]
```diff
@@ -2,44 +2,55 @@ package main
...
 
-	cmd := exec.Command("go", "build", ".")
-	err := cmd.Run()
+	tmp, err := os.MkdirTemp("", "go-python-embedding-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmp)
+
+	cmd := exec.Command("go", "build", "-o", filepath.Join(tmp, "exe"), ".")
```
Comment (sbinet) [inline examples/embedding/main_test.go:24]: little windows snowflake :)
```suggestion
	exe := filepath.Join(tmp, "out.exe")
	cmd := exec.Command("go", "build", "-o", exe, ".")
```

File: examples/embedding/main_test.go [hunk 9]
```diff
@@ -2,44 +2,55 @@ package main
...
+	err = cmd.Run()
 	if err != nil {
 		t.Fatalf("failed to compile embedding example: %v", err)
 	}
 
 	out := new(bytes.Buffer)
-	cmd = exec.Command("./embedding", "mylib-demo.py")
+	cmd = exec.Command(filepath.Join(tmp, "exe"), "mylib-demo.py")
```
Comment (sbinet) [inline examples/embedding/main_test.go:31]:
```suggestion
	cmd = exec.Command(exe, "mylib-demo.py")
```

**Comment by drew-512** [comment]
yay!

**Review by sbinet — APPROVED** [review]
LGTM

**Pull request merged**